### PR TITLE
DM-25991: Update ap_pipe plotting tools

### PR DESCRIPTION
**Pull request by mrawls**
Nearly done here, a few more plotting functions coming, rebase pending, comments welcome

### CODE REVIEW

**Comment by kherner** [comment]
A few comments on the changes to coaddAnalysis.py:

I would extend the list of checks at the beginning of each function to print a useful error message if things like skymap and collection aren't provided when running in gen3. For example near the beginning of the getPatchConstituents function 

>   elif gen == 'gen3':
>       butler = dafButler.Butler(repo, collections=collections)
>       skymap = butler.get('skyMap', instrument=instrument, collections='skymaps', skymap=skymapName)

I would add something like (apologies if the indentation doesn't come through)

`elif gen =='gen3':`
`    if not collections:`
`        raise ValueError('A non-empty list of collections is required in gen3.')`
`    if not skymap:`
`        raise ValueError('A skymap string is required in gen3.')`
`    if not instrument:`
`        raise ValueError('An instrument string is required in gen3.')`

I would also add the same in mosaicCoadd when first trying to set up the Butlers. 

Also in mosaicCoadd I'd add the same check for "gen" being set to either gen2 or gen3 that you have in the getPatchConstituents function.

**Comment by kherner** [comment]
Comment #2: Since I have vowed to work more or less exclusively in gen3, I would advocate for making it the default, though that is probably (almost certainly?) the more disruptive course of action. I won't go to the mattresses over it, certainly.

**Comment by kherner** [comment]
Comment #3: When actually getting the coadd array in mosaicCoadd (say lines 370 - 386) in the current draft I would suggest wrapping all that in a try block (at least in Gen 3), because there's currently nothing stopping someone from calling the function with a patch list that contains entries that might be valid patches, but don't have any data. 

> if patchIndex in patchIndicesToLoad:
> ...
>                else:
>                   ax.text(.3, .3, patchIndex)  

Would become something like

> if patchIndex in patchIndicesToLoad:
>        try:
> ...
> except:
> (indent) print('Error getting data for patch {}.'.format(patchIndex))
> (indent) ax.text(.3, .3, patchIndex)
>                else:
>                   ax.text(.3, .3, patchIndex)

**Comment by mrawls** [comment]
This PR includes addressing @kherner's comments, though I have kept gen2 as default for preexisting plotting functions.

File: apdbPlots.py [hunk 1]
```diff
@@ -269,8 +282,10 @@ def getCcdAndVisitSizeOnSky(repo, sourceTable, gen='gen2', instrument='DECam', c
         Either 'gen2' or 'gen3'
     instrument : `str`, optional
         Default is 'DECam', used with gen3 butler only
-    collections : `list`, optional
+    collections : `list` or `str`, optional
         Must be provided for gen3 to load the camera properly
+    visit : `int` or None, optional
+    detector : `int` or None, optional
```
Comment (isullivan) [inline apdbPlots.py:288]: Please add one-line descriptions.

File: apdbPlots.py [hunk 2]
```diff
@@ -699,7 +768,178 @@ def plotDiaSourcesOnSkyGrid(repo, sourceTable, title=None, color='C0'):
...
         ax.get_xaxis().set_visible(False)
         ax.get_yaxis().set_visible(False)
-    fig.tight_layout()
+    fig.tight_layout(rect=[0, 0.03, 1, 0.95])
     plt.subplots_adjust(wspace=0)
     if title:
         fig.suptitle(title)
+
+
+def plotFlagHist(sourceTable, title=None, defaultBadFlags=True, colorList=None):
```
Comment (isullivan) [inline apdbPlots.py:777]: Would it be possible to pass in a list of bad flag names, and programmatically color those red and the rest blue?

File: apdbPlots.py [hunk 3]
```diff
@@ -699,7 +768,178 @@ def plotDiaSourcesOnSkyGrid(repo, sourceTable, title=None, color='C0'):
...
+            color = 'C2'
+        elif band == 'r':
+            color = 'C1'
+        elif band == 'i':
+            color = 'C3'
+        elif band == 'z':
+            color = 'C5'
+        else:
+            color = 'k'  # should be y
+        butler = dafButler.Butler(repo)
```
Comment (isullivan) [inline apdbPlots.py:923]: Shouldn't the `butler` definition happen outside the `for` loop?

File: coaddAnalysis.py [hunk 4]
```diff
@@ -32,18 +46,38 @@ def getPatchConstituents(repo, band='g', printConstituents=False, verbose=False)
...
         The number of visits that contributed to each patch.
     """
-    butler = dafPersist.Butler(repo)
-    skymap = butler.get('deepCoadd_skyMap', dataId={'filter': band})
-    tractInfo = skymap[0]  # implicitly assumes there is just 1 tract
-    everyPatchList = [str(patch.getIndex()[0]) + ',' + str(patch.getIndex()[1]) for patch in tractInfo]
+    if gen == 'gen2':
+        butler = dafPersist.Butler(repo)
+        skymap = butler.get('deepCoadd_skyMap', dataId={'filter': band})
+        tract = skymap[0]  # implicitly assumes there is just 1 tract
```
Comment (isullivan) [inline coaddAnalysis.py:52]: Probably irrelevant, but couldn't this be `tract = skymap[tractIndex]`?

File: coaddAnalysis.py [hunk 5]
```diff
@@ -32,18 +46,38 @@ def getPatchConstituents(repo, band='g', printConstituents=False, verbose=False)
...
+        everyPatchList = [str(patch.getIndex()[0]) + ',' + str(patch.getIndex()[1]) for patch in tract]
+    elif gen == 'gen3':
```
Comment (isullivan) [inline coaddAnalysis.py:54]: In a previous function, you let gen3 be the default by using an `else` for cases like this. Would it make sense to do that here (and a few lines down)?

File: coaddAnalysis.py [hunk 6]
```diff
@@ -162,10 +207,15 @@ def percent(values, p=0.5):
     axes.fill(tractRa, tractDec, fill=True, color="black", lw=1, linestyle='solid',
               edgecolor='k', alpha=0.2)
     for ip, patch in enumerate(tractInfo):
-        patchIndexStr = str(patch.getIndex()[0]) + "," + str(patch.getIndex()[1])
+        if gen == 'gen2':
+            patchIndexStr = str(patch.getIndex()[0]) + "," + str(patch.getIndex()[1])
+            patchWithData = patchIndexStr
+        elif gen == 'gen3':
```
Comment (isullivan) [inline coaddAnalysis.py:213]: Should this just be an `else`?

File: coaddAnalysis.py [hunk 7]
```diff
@@ -190,21 +240,24 @@ def percent(values, p=0.5):
...
-    butler : `lsst.daf.base.Butler`
-        A Butler instance to retrieve data from a repository.
-    patch_list : `list` of `str`
-        A list of the patches containing images to mosaic.
+    repo : `str`
+        The path to the data repository.
+    patch_list : `list`
+        A list of the patch indices containing images to mosaic.
+        In gen2, list elements will be `str` e.g. '(1, 2)'
+        In gen3, list elements will be `float` (sequential patch indices)
```
Comment (isullivan) [inline coaddAnalysis.py:255]: Are they actually floats, or ints?

File: coaddAnalysis.py [hunk 8]
```diff
@@ -226,56 +279,141 @@ def mosaicCoadd(butler, patch_list, band='g', tract=0, ref_patch=None, sampling=
...
+                ax = plt.subplot(gs1[figIdx])
+                patchIndex = (x, ny - y - 1)
+                if patchIndex in patchIndicesToLoad:
+                    sequentialPatchIndex = [key for key, value in indexmap.items() if value == patchIndex][0]
+                    dataId = {'band': band, 'tract': tractIndex, 'patch': int(sequentialPatchIndex),
+                              'instrument': instrument, 'skymap': skymapName}
+                    try:
+                        coadd = butler.get(coaddName, dataId=dataId)
+                    except LookupError:
+                        print('Failed to retrieve data for patch {0}'.format(patchIndex))
```
Comment (isullivan) [inline coaddAnalysis.py:389]: `print(f'Failed to retrieve data for patch {patchIndex}'.)`?

File: plotLightcurve.py [hunk 9]
```diff
@@ -1,152 +1,109 @@
...
+from matplotlib import cm
 import pandas as pd
-import sqlite3
-from astropy.visualization import (ZScaleInterval, SqrtStretch, ImageNormalize)
+import astropy.visualization as aviz
 
 import lsst.daf.persistence as dafPersist
-import lsst.afw.display as afwDisplay
+# import lsst.daf.butler as dafButler
+# import lsst.afw.display as afwDisplay
```
Comment (isullivan) [inline plotLightcurve.py:10]: Remove these commented-out imports

File: plotLightcurve.py [hunk 10]
```diff
@@ -1,152 +1,109 @@
...
-            return True
-        else:
-            return False
-    except NameError:
-        return False
-
-
-def loadApdbObjects(repo, dbName='association.db', filter=''):
-    """Load select DIAObject columns from a APDB into a pandas dataframe.
+def loadSomeApdbSources(dbName, diaObjectId, dbType='sqlite', schema=None):
```
Comment (isullivan) [inline plotLightcurve.py:21]: Could this be called `loadApdbSources` or `loadSelectedApdbSources`? `Some` sounds like they could be random.

File: plotLightcurve.py [hunk 11]
```diff
@@ -1,152 +1,109 @@
...
+                                  "midPointTai", "apFlux", "psFlux", "apFluxErr", \
+                                  "psFluxErr", "totFlux", "totFluxErr", "x", "y", \
+                                  "ixxPSF", "iyyPSF", "ixyPSF", "flags" from {0}; \
+                                  where "diaObjectId" = {1};'.format(tables['src'], diaObjectId), connection)
+    return srcTable
+
 
-    # These are the tables available in the APDB
-    tables = {'obj': 'DiaObject', 'src': 'DiaSource', 'ccd': 'CcdVisit'}
+def loadExposures(butler, dataId, collections):
```
Comment (isullivan) [inline plotLightcurve.py:54]: This needs a full docstring.

File: plotLightcurve.py [hunk 12]
```diff
@@ -1,152 +1,109 @@
...
+    return science, difference, template
 
-    # Only get objects with validityEnd NULL because that means they are still valid
-    objTable = pd.read_sql_query('select diaObjectId, ra, decl, nDiaSources, \
-                                  gPSFluxMean, validityEnd, flags from {0} \
-                                  where validityEnd is NULL \
-                                  {1};'.format(tables['obj'], filter), connection)
-    return objTable
 
+def retrieveCutouts(butler, dataId, collections, center, size=lsst.geom.Extent2I(30, 30)):
```
Comment (isullivan) [inline plotLightcurve.py:64]: This needs a full docstring.

File: plotLightcurve.py [hunk 13]
```diff
@@ -1,152 +1,109 @@
...
-def loadApdbSources(dbPath, obj):
-    """Load select DIAObject columns from a APDB into a pandas dataframe.
 
-    Parameters
-    ----------
-    repo : `str`
-        Path to the APDB.
-    obj : `int`
-        DIA Object for which we want to retrieve constituent DIA Sources.
+def plotCutout(scienceCutout, differenceCutout, templateCutout, output=None):
```
Comment (isullivan) [inline plotLightcurve.py:75]: This needs a full docstring.

**Review by isullivan — APPROVED** [review]
I focused on the plotting functions, and have not reviewed the notebooks. There is a lot of changed code so I did have a few requested changes, but overall it looks like a nice improvement.

File: apdbPlots.py [hunk 14]
```diff
@@ -41,20 +41,31 @@ def addVisitCcdToSrcTable(sourceTable, cam='decam'):
         Pandas dataframe with DIA Sources from an APDB.
     cam : `str`, one of either 'decam' or 'hsc', default is 'decam'
         Needed to properly add the "ccd" and "visit" columns to the sourceTable.
+    gen : `str`, optional
+        Either 'gen2' or 'gen3'
+    instrument : `str`, optional
+        Default is 'DECam', used with gen3 butler only
```
[the remark this answers is not in the document]
Comment (mrawls) [reply apdbPlots.py:47]: Actually Jim has fixed this so the gen3 version should work for all instruments! Updated accordingly.

File: apdbPlots.py (hunk 2 above)
Comment (mrawls) [reply apdbPlots.py:777]: Good point, that would be much more useful

File: coaddAnalysis.py (hunk 7 above)
Comment (mrawls) [reply coaddAnalysis.py:255]: They're freaking floats

File: coaddAnalysis.py (hunk 5 above)
Comment (mrawls) [reply coaddAnalysis.py:54]: Fixed several of these, good catch

File: plotLightcurve.py (hunk 10 above)
Comment (mrawls) [reply plotLightcurve.py:21]: `loadAnyOldApdbSources` ... haha fair enough

**Review by isullivan — CHANGES_REQUESTED** [review]
In `DM-28555-Investigate-diaPipe-changes.ipynb` please expand the description/explanation of the notebook at the top. A brief comment summarizing the problem and stating that it appears to be fixed with the fix from DM-28555 should be sufficient.

In `Spring-2021-diffim-sprint-first-big-run.ipynb`, it looks like we do not have the correct flags set to filter out the *many* spurious sources around bright objects. Could you investigate the available flag options further, to see if any of our existing flags do a good job of cleaning up the diaSource scatter plots? Based on the flag histogram, I would suggest starting with `slot_Shape_flag_unweightedBad`, `base_PixelFlags_flag_interpolatedCenter`, `base_PixelFlags_flag_cr`, and `base_PixelFlags_flag_interpolated`. If it is not too much work, a scatter plot for each type of flag in that histogram showing the distribution of sources with each flag set would be very interesting, and would probably tell us which flags we should be using.

**Pull request merged**